[MP Rewrite Part 2] Introduce the Trajectory abstraction

kylestach opened the pull request:
A Trajectory is a piecewise-cubic function interpolating between a
vector of time-indexed RobotInstant waypoints.

This is not expected to pass CI.

kylestach committed:
[MP Rewrite Part 2] Introduce the Trajectory abstraction

A Trajectory is a piecewise-cubic function interpolating between a
vector of time-indexed RobotInstant waypoints.

This is not expected to pass CI.

kylestach committed:
Merge branch 'mp-rewrite/ball-state' into mp-rewrite/trajectory

kylestach committed:
Add motion control changes for following Planning::Trajectory

kylestach committed:
Merge branch 'mp-rewrite/ball-state' into mp-rewrite/trajectory

oswinso commented on common/Utils.hpp:
```diff
@@ -61,10 +61,7 @@ inline void debugThrowIf(const std::string& string, bool condition) {
-    while (a < -M_PI) a += 2.0 * M_PI;
-    while (a > M_PI) a -= 2.0 * M_PI;
-    return a;
+    return std::atan2(std::sin(a), std::cos(a));
```
Why this change?

kylestach replied:
O(1) instead of O(n), lmao. This should really just be an `std::remainder` though.

oswinso replied:
I'm pretty sure with the old one the `while` acts more like an `if` though... but I think it's fine. Just a bit worried about the performance of calling `sin`, `cos` and `atan2`

kylestach replied:
Remainder is preferable to any of the above. I was meaning to swap that out but this was the first change I thought of (I just really don't like the idea that we could take an unbounded amount of time in this stupid little function, especially if something hits infinity or a really really large number).

kylestach replied:
Alright `std::remainder` has the correct behavior, I'll swap out to that.

kylestach replied:
I didn't catch that we were doing a remainder beforehand, my bad. That should already be correct (always between -pi and pi), and we shouldn't even have needed the loops.

oswinso commented on soccer/motion/MotionControl.cpp:
```diff
@@ -102,7 +98,7 @@ void MotionControl::run(const RobotState& state,
     // Use default constraints. Planning should be in charge of enforcing
-    // constraints on the path, here we just follow it.
+    // constraints on the trajectory, here we just follow it.
     // TODO(Kyle): Use this robot's constraints here.
```
I know this isn't part of this PR but issue number plz

kylestach replied:
#1500

oswinso commented on soccer/planning/Instant.hpp:
```diff
@@ -0,0 +1,94 @@
+namespace Planning {
+
+/**
+ * This class represents a robot's motion "state" at a given time, including
```
brief

oswinso commented on soccer/planning/Instant.hpp:
```diff
@@ -0,0 +1,94 @@
+};
+
+/**
+ * Represents the current state of a robot in a planned trajectory.
```
brief

oswinso commented on soccer/planning/TrajectoryUtils.hpp:
```diff
@@ -0,0 +1,42 @@
+
+namespace Planning {
+
+// TODO(Kyle): both of these functions should be able to be abstracted under a
+//  single interface templated on either a shape set or a collection of dynamic
+//  obstacles.
```
Issue number pls

kylestach replied:
#1501

oswinso commented on soccer/planning/Trajectory.cpp:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    RobotInstantSequence result_instants;
+    // TODO(Kyle): Reserve space in result_instants.
```
Issue num

oswinso reviewed (changes requested):
Preliminary changes

kylestach committed:
Merge branch 'mp-rewrite/ball-state' into mp-rewrite/trajectory

oswinso commented on soccer/planning/Trajectory.hpp:
```diff
@@ -0,0 +1,420 @@
+     * @details Intended for use when editing a trajectory in-place; for other
+     *  cases it may be simpler to use the iterator or Cursor interfaces.
+     */
+    [[nodiscard]] auto& instant_at(int i) { return instants_.at(i); }
```
Why the `auto` in these functions?

oswinso commented on soccer/planning/TrajectoryUtils.cpp:
```diff
@@ -0,0 +1,141 @@
+        RobotInstant instant = cursor.value();
+
+        // Calculate the list of hits.
+        // TODO(Kyle): We could save a lot of allocation and smart pointer
+        //  reference counting overhead by making hitSet return a custom
+        //  iterator type. Then, we could use that to construct a set at the
+        //  beginning but just iterate over it in standard fashion here.
```
issue number

oswinso reviewed (changes requested):
more things

kylestach committed:
Address review comments

kylestach committed:
Change fixAngleRadians to be more efficient

egordon9dev commented on soccer/planning/Trajectory.cpp:
```diff
@@ -0,0 +1,344 @@
+    ScaleDuration(final_duration, begin_time());
+}
+
+void Trajectory::ScaleDuration(RJ::Seconds final_duration,
```
I forgot where this is used, but this will be a bit slower than the old version. It's probably fine though

kylestach replied:
It's not actually used anywhere, I think. Yeah it'll be O(n) instead of O(1) but it's super infrequently used.

egordon9dev replied:
Yeah that's fine. Trajectories are smaller now. I didn't realize that when making this comment so this is not really an issue.

egordon9dev commented on soccer/planning/Trajectory.hpp:
```diff
@@ -0,0 +1,422 @@
+     * @brief Create a trajectory from several "instants", each with a pose,
+     * velocity, and timestamp. This
+     */
+    explicit Trajectory(RobotInstantSequence instants)
```
should use an r-value reference or a smart pointer in the function parameter. I think this makes a copy

kylestach replied:
To pass into this you can do `Trajectory(std::move(instants_arg))` and it will move `instants_arg` into the `instants` argument (call the move constructor for the parameter. This is generally preferred because you can pass either a const reference or an rvalue.

egordon9dev commented on soccer/planning/Trajectory.hpp:
```diff
@@ -0,0 +1,422 @@
+     *
+     * @details Is able to move out of the first trajectory, but not the second.
+     */
+    Trajectory(Trajectory a, const Trajectory& b);
```
Maybe have an r-value reference / smart pointer version of this function, so we can move out of the first Trajectory. Here we are copying both.

kylestach replied:
See above

kylestach committed:
Merge branch 'mp-rewrite/ball-state' into mp-rewrite/trajectory

JNeiger commented on soccer/planning/Trajectory.cpp:
```diff
@@ -0,0 +1,344 @@
+
+RobotInstant Trajectory::Cursor::value() const {
+    RobotInstant instant_0 = *iterator_;
+    if (instant_0.stamp == time_) {
```
`*iterator_.stamp` and `time_` are only equal if it's the last instant in the vector right?

kylestach replied:
Yes. This check is necessary to avoid dereferencing the `end()` iterator in the case that we're pointing to the end.

JNeiger commented on soccer/planning/Trajectory.hpp:
```diff
@@ -0,0 +1,422 @@
+     * @brief Allows seeking to arbitrary positions in a Trajectory and
+     *  incrementing by arbitrary steps.
+     */
+    class Cursor {
```
This header and source file are pretty large, would it be simple to split them up?

kylestach replied:
It would be a bit annoying, `Cursor` and `Trajectory::Cursor` are pretty much interdependent.

JNeiger replied:
Nevermind then

JNeiger commented on soccer/planning/TrajectoryUtils.cpp:
```diff
@@ -0,0 +1,141 @@
+
+    // Limit iterations to 400. This will continue to operate at dt = 0.05 until
+    // we hit a 20 second trajectory. If our trajectory is longer than that,
+    // something is probably wrong.
```
You should specify that the dt increases after 20 seconds as well

JNeiger commented on common/Geometry2d/Pose.hpp:
```diff
@@ -128,12 +128,38 @@ class Pose {
+     * Inequality comparison operation.
+     */
+    bool operator!=(const Pose& other) const {
+        return _position != other._position || _heading != other._heading;
```
Since you have the equality operator, you can just do `!(*this == other)`. This reduces copy paste errors

JNeiger reviewed (commented):
Just a few things here and there

kylestach committed:
Address review comments

JNeiger commented on common/Geometry2d/Pose.hpp:
```diff
@@ -289,12 +313,38 @@ class Twist {
+    /**
+     * Inequality comparison operation.
+     */
+    bool operator!=(const Twist& other) const {
```
`!=` oeprator that you can swap to `!( == )`

JNeiger commented on soccer/planning/Instant.hpp:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * Inequality comparison operator.
+     */
+    bool operator!=(const RobotInstant& other) const {
```
`!=` operator that you can swap to `!( == )`

JNeiger reviewed (commented):
Missed two `!=` operators. LGTM after that

kylestach committed:
Tidy up != operators (use == in defn.)

oswinso reviewed (commented):
lgtm

oswinso reviewed (approved):
lgtm

kylestach committed:
[MP Rewrite Part 3] Add PlanRequest and the planner node. (#1495)

* Add PlanRequest and the planner node.

This also overhauls the MotionCommand structure.

* Remove dead code from Processor, replace with TODO

* Make path target goal a linear motion instant

* Address review comments

[MP Rewrite Part 4] Motion Planning Primitives (#1496)

Primitives for motion planning including angle planning, RRT, velocity
profiling, Bezier fitting, and replanning. This is anticipated to be the
largest portion of the motion planning rewrite.

This is not expected to pass CI.

* Remove disabled for-loop in Processor

* Fix TODOs and docs in planning primitives

* Address review comments

* Make off-path threshold configurable

[MP Rewrite Part 5] Port previous planners to new framework (#1497)

Most of these are not functional changes from the old planners, but
rather simple ports to the new primitives. The biggest changes are that
we now use `Replanner` static functions and manually manage our own
"previous trajectory" objects instead of having "behavior-level"
planners (i.e. settle) own instances of an RRT planner and interacting
with it through plan requests.

This is the final PR in the rewrite, and it should pass CI. At the point
where it passes CI and all PRs are ready for merge, each PR should be
merged into its parent in sequence.

* Address review comments

* Move applyLowPassFilter into common, add TODOs

* Enable escape-obstacles planner

[MP Rewrite Part 6] Cleanup and fixes (#1512)

* Cleanup and fixes to make code run properly

* Fix test compilation and disable a trapezoidal motion profile edge case.

Some planner tests are still failing. This will be addressed in a later
commit.

* Prettify

* Make e2e planner tests pass.

Because they generate random scenarios, and the RRT planner is not
guaranteed to converge in a finite number of iterations (and the Bezier
approximation may be incorrect), these tests are currently flaky. That
should be fixed before this is merged to staging.

* Tidy fixes

* Linting

* Formatting

* Remove deleted file from CMakeLists, remove reuse condition from settle

* Don't replan in Collect::control

* Optimizations to geometry methods

* Reset and regenerate sim config

This deletes unnecessary variables and moves planning-related variables
into PathPlanner

* Formatting

* Add TODO in EscapeObstacles

* Tidy fixes

* Revert inconsistent clang-tidy recommended fix

* Temporarily disable flaky planner test

* Add NOLINTNEXTLINE

* Remove old sim config and add TODOs for trajectory utils
